test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,443 @@
+				if progTestType != spec.Type {
+					if progTestType == ebpf.UnspecifiedProgram {
+						progTestType = spec.Type
+						continue
+					}
+					if spec.Type == ebpf.UnspecifiedProgram {
+						continue
+					}
+
+					t.Fatalf(
+						"File '%s' contains both '%s' and '%s' program types, "+
+							"only one program type per ELF file allowed:",
+						elfPath,
+						progTestType,
+						spec.Type,
+					)
+				}
+			}
+
+			if progTestType == ebpf.UnspecifiedProgram {
+				t.Fatalf("File '%s' only contains unspecified program types", elfPath)
+			}
+
+			// Give all tail call programs the same program type as the test programs
+			for _, spec := range spec.Programs {
+				spec.Type = progTestType
+			}
+
+			coll, err := ebpf.NewCollectionWithOptions(spec, ebpf.CollectionOptions{
+				Programs: ebpf.ProgramOptions{
+					LogSize: 1 << 20,
+				},
+				Maps: ebpf.MapOptions{
+					PinPath: pinPath,
+				},
+			})
+			if err != nil {
+				t.Fatal("new coll:", err)
+			}
+			defer coll.Close()
+
+			err = loadCallsMap(spec, coll)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			testNameToPrograms := make(map[string]programSet)
+
+			for progName, spec := range spec.Programs {
+				match := checkProgRegex.FindStringSubmatch(spec.SectionName)
+				if len(match) == 0 {
+					continue
+				}
+
+				progs := testNameToPrograms[match[1]]
+				if match[2] == "setup" {
+					progs.setupProg = coll.Programs[progName]
+				}
+				if match[2] == "check" {
+					progs.checkProg = coll.Programs[progName]
+				}
+				testNameToPrograms[match[1]] = progs
+			}
+
+			for progName, set := range testNameToPrograms {
+				if set.checkProg == nil {
+					t.Fatalf(
+						"File '%s' contains a setup program in section '%s' but no check program.",
+						elfPath,
+						spec.Programs[progName].SectionName,
+					)
+				}
+			}
+
+			// Collect debug events and add them as logs of the main test
+			var globalLogReader *perf.Reader
+			if coll.Maps["test_cilium_events"] != nil {
+				globalLogReader, err = perf.NewReader(coll.Maps["test_cilium_events"], 4096*16)
+				if err != nil {
+					t.Fatalf("new global log reader: %s", err.Error())
+				}
+
+				linkCache := link.NewLinkCache()
+
+				// Closes as soon as the goroutine has started
+				started := make(chan struct{})
+
+				go func() {
+					defer globalLogReader.Close()
+
+					close(started)
+					for {
+						rec, err := globalLogReader.Read()
+						if err != nil {
+							return
+						}
+
+						dm := monitor.DebugMsg{}
+						reader := bytes.NewReader(rec.RawSample)
+						if err := binary.Read(reader, byteorder.Native, &dm); err != nil {
+							return
+						}
+
+						t.Log(dm.Message(linkCache))
+					}
+				}()
+				<-started
+			}
+
+			for name, progs := range testNameToPrograms {
+				t.Run(name, subTest(progs, coll.Maps[suiteResultMap]))
+			}
+
+			if globalLogReader != nil {
+				// Give the global log buf some time to empty
+				time.Sleep(50 * time.Millisecond)
```

Comment:
:thinking:

Reply:
cilium/ebpf currently doesn't provide this, though I agree it should, and I believe it is possible, I will see if I can fix that. Would it be acceptable to leave it like this and fix it as soon as a flush feature has been upstreamed in cilium/ebpf? (if that takes longer than this PR)

Reply:
Maybe a question first, what's the consequences if 50 Milliseconds is not enough time to flush the log? Can we easily detect that problem and react to it?